sympy/integrals/integrals.py

```diff
@@ -1242,6 +1242,105 @@ def _sage_(self):
+        .. [2] http://mathworld.wolfram.com/CauchyPrincipalValue.html
+        """
+        from sympy.calculus import singularities
+
+        if len(self.limits) != 1 or len(list(self.limits[0])) != 3:
+            raise ValueError("You need to insert a variable, lower_limit, and upper_limit correctly to calculate "
+                             "cauchy's principal value")
+
+        x, a, b = self.limits[0]
+
+        if not(a.is_comparable and b.is_comparable and a <= b):
+            raise ValueError("The lower_limit must be smaller than or equal to the upper_limit to calculate "
+                             "cauchy's principal value. Also, a and b need to be comparable.")
+
+        if a == b:
+            return 0
+        r = Dummy('r')
+        u = Dummy('u')
+
+        singularities_list = [s for s in singularities(self.function, x) if s.is_comparable and a <= s <= b]
+
+        for i in singularities_list:
+            if ((i == b) or (i == a)):
+                raise ValueError(
+                    'The principal value is not defined in the given interval due to singularity at %d.' % (i))
+
+        f = self.function
+        F = integrate(f, x)
+
+        if F.has(Integral):
+            return self
+        if len(singularities_list) == 0:
+            if a is -oo and b is oo:
+                I = limit(F - F.subs(x, -x), x, oo)
+                return I
+            else:
+                I = limit(F, x, b, '-') - limit(F, x, a, '+')
+                return I
+
+        if len(singularities_list) == 1 and not(a is -oo and b is oo):
+            I = limit(((F.subs(x, singularities_list[0] - r)) - F.subs(x, singularities_list[0] + r)), r, 0,
+                      '+') - limit(F, x, a, '+') + limit(F, x, b, '-')
+            return I
+
+
+        else:
+            if a is -oo and b is oo:
+                Iinf = 0
+                I_lim_subs = limit(F.subs(x, u) - F.subs(x, -u), u, oo)
+                for singular_element in range(len(singularities_list)):
+                    Iinf = Iinf + limit(((F.subs(x, singularities_list[singular_element] - r)) - F.subs(x,
+                                                                                                        singularities_list[
+                                                                                                            singular_element] + r)),
+                                        r, 0, '+')
+                return Iinf + I_lim_subs
+            else:
+                Ib = 0
+                Ia = F.subs(x, singularities_list[0] - r) - limit(F, x, a, '+')
+                for singular_element in range(1, len(singularities_list) - 1):
+                    Ib = Ib + (F.subs(x, singularities_list[singular_element] - r) - F.subs(x, singularities_list[
+                        singular_element - 1] + r))
+                I1 = limit((limit(F, x, b, '-') - (F.subs(x, singularities_list[len(singularities_list) - 1] + r))) + (
```

Comment:
`limit((limit(F, x, b, '-')` can be added to `Ia` in place of `F.subs(x, singularities_list[0] - r)`.

Reply:
@jksuom  Again, that wouldn't have any advantage over what is currently implemented. The current code will be easier to understand for an outside reader since `I1` sums up all the integrals i.e. the first integral ( given by `Ia` ), the other remaining integrals except the last two ( given by `for` loop ) and then `I1` adds all of them to the last two integrals and implements `limit`. 
adding `limit((limit(F, x, b, '-')` to `Ia` would therefore make the code confusing and wouldn't give any advantage.

Reply:
@jksuom  Don't you think so, Sir?

Reply:
Note that the singularities are not in any particular order.

Reply:
I think that it would be confusing to combine `limit((limit(F, x, b, '-')` with `singularities_list[len(singularities_list) - 1]` which could be any one of the singularities.

Reply:
I think that it should not be hard to understand the code that I proposed:

    for s in singularities_list:
        I += F.subs(x, s - t) - F.subs(x, s + t)

which means that a piece around the singularity is omitted from the original `I` from `a` to `b`.
The limit form should not be hard to understand either: 

    for s in singularities_list:
        I += limit(F.subs(x, s - t) - F.subs(x, s + t), t, 0, '+')

That might be easier to compute in some cases.

The ordering of the singularities does not matter.

Reply:
@jksuom 
>Note that the singularities are not in any particular order.

Oh! I did not know that. I believed that they are presented in increasing order. Sure enough, I'll implement your review. I'll be making another commit soon.